### PR TITLE
Include unpublishable path dependencies of the root package for publishing

### PR DESCRIPTION
**Pull request by iology**
ping #4726 

Before this patch, all path deps must be published before the root crate is published to the registry.

This patch allow `cargo package` to include files of path dependencies in the generated package if the dependencies have the manifest key "package.publish" set to false, which means the crate cannot be registered as a sharable crate by itself at a registry, but still allowed to be published along with the root crate.

### CODE REVIEW

**Comment by matklad** [comment]
cc @rust-lang/cargo 

So basically this allows one to have "private internal"  crates, which can't be used outside of the current package?

I personally would really love to have this feature: it's important to be able to create an appropriate physical layout in terms of compilation units, without exposing the details to the consumer of the package. But I am quite surprised by the fact that there's no **huge** demand for this feature, so perhaps it's not that important? This also is related to [rfc1997](https://github.com/rust-lang/rfcs/blob/master/text/1977-public-private-dependencies.md) about private dependencies.

**Comment by matklad** [comment]
Thanks for the pull-request @iology!

I think this actually can work like this, without changing the fundamental property that Cargo package contains a single library, because all other "libraries" are strictly private and not visible from outside! But we probably need an RFC for this, because there may be some non-obvious ramifications. For example, I think that for such "internal" packages a lot of fields from Cargo.toml, most importantly `version`, are irrelevant. So perhaps such "private packages" should have more first-class support in Cargo?

**Comment by iology** [comment]
@matklad You are right. There may still be some problems about rustdoc and I haven't added more tests. And I must say that I'm still a newbie of Rust, have only made some small programs. So your opinions are definitely needed! You can take this patch as just a proof of concept!

**Comment by joshtriplett** [comment]
I do agree that this needs an RFC, to specify the semantics here. That's not a hard process, and there are many people who can help you with it. It's a great way to get started with contributing to Rust.

**Comment by iology** [comment]
Is there any RFC about this topic? `package.publish=false` may carry too much semantics, so another attribute may be needed to override `package.publish`.

File: src/cargo/util/toml/mod.rs [hunk 1]
```diff
@@ -905,6 +937,22 @@ impl TomlDependency {
             cx.warnings.push(msg);
         }
 
+        if details.git.is_none() && details.registry.is_none() &&
+            details.path.is_some() {
+            match kind {
+                // `cargo [build|test|bench]` are not affected.
+                Some(Kind::Normal) => {
```
Comment (iology) [inline src/cargo/util/toml/mod.rs]: Bug: currently we cannot know if the manifest is loaded for *cargo install* or *cargo package*. `kind` is always None for normal dependencies.

File: tests/registry.rs [hunk 2]
```diff
@@ -252,26 +252,101 @@ fn package_with_path_deps() {
         .build();
 
     assert_that(p.cargo("package").arg("-v"),
-                execs().with_status(101).with_stderr_contains("\
```
Comment (iology) [inline tests/registry.rs]: Shoud be the original error. This patch is not to change any behavior when `package.publish=true`.

File: src/cargo/util/toml/mod.rs (hunk 1 above)
Comment (iology) [reply src/cargo/util/toml/mod.rs]: In fact, I don't know why I'm checking this here. The original code doesn't have this logic too. :P

**Comment by pornel** [comment]
This feature sounds very interesting to me. For example I have a crate `c3` that uses a sub crate to contain pains of building and patching clang+llvm (https://crates.io/crates/c3_clang_extensions), so I had to publish it, but it's not useful outside of its parent.

**Comment by alexcrichton** [comment]
I'm going to close this for now in favor of the RFC @matklad linked to

**Pull request closed without merging**